### PR TITLE
Log aggregation

### PR DESCRIPTION
**Pull request by paulschwarzenberger**
S3 and Lambda for log data aggregation

### CODE REVIEW

File: s3/variables.tf [hunk 1]
```diff
@@ -77,3 +77,23 @@ variable "force_destroy" {
   description = "destroys S3 bucket on terraform destroy, even if there are files inside the bucket"
   default     = true
 }
+
+variable "sns_topic_arn" {
```
Comment (TomJKing) [inline s3/variables.tf:81]: Should these variable name be more specific to log aggregation? Just thinking if we need to add additional sns topics in the future to S3 buckets

File: lambda/functions/log-data/lambda_function.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+import boto3
+import ast
+import json
+import os
+from urllib.parse import unquote
+print('Loading function')
+
+def lambda_handler(event, context):
```
Comment (TomJKing) [inline lambda/functions/log-data/lambda_function.py:8]: Was terraform fmt applied to this file?

File: sns/main.tf [hunk 3]
```diff
@@ -0,0 +1,12 @@
+resource "aws_sns_topic" "log_aggregation" {
```
Comment (TomJKing) [inline sns/main.tf:1]: For the sns topic module, is the idea to add additional sns topics to this main.tf?

File: sns/outputs.tf [hunk 4]
```diff
@@ -0,0 +1,3 @@
+output "sns_arn" {
```
Comment (TomJKing) [inline sns/outputs.tf:1]: Should this output name be more specific to log aggregation> Just thinking if further sns topics are added in the future

File: s3/locals.tf [hunk 5]
```diff
@@ -4,4 +4,12 @@ locals {
   standard_bucket_name = "${var.project}-${var.function}-${local.environment}"
   global_bucket_name   = "${var.project}-${var.function}"
   bucket_name          = var.environment_suffix == true ? local.standard_bucket_name : local.global_bucket_name
+  sns_topic_arn        = var.sns_topic_arn == "" ? "arn:aws:sns:${var.sns_topic_region}:${data.aws_caller_identity.current.account_id}:${var.project}-logs-${local.environment}" : var.sns_topic_arn
```
Comment (TomJKing) [inline s3/locals.tf:7]: Should this be named "log_aggregation_sns_topic_arn" as likely will need to add another sns topic for the backend checks

File: s3/variables.tf (hunk 1 above)
Comment (paulschwarzenberger) [reply s3/variables.tf:91]: Thanks, separated SNS variables into log SNS variables, for log bucket, and the SNS variables for main S3 bucket

File: lambda/functions/log-data/lambda_function.py (hunk 2 above)
Comment (paulschwarzenberger) [reply lambda/functions/log-data/lambda_function.py:8]: no, as this is a python file

File: sns/main.tf (hunk 3 above)
Comment (paulschwarzenberger) [reply sns/main.tf:1]: The idea is to separately call the SNS module for each new SNS topic, with appropriate variables - the main.tf in the modules only needs to contain a single SNS topic.

File: s3/locals.tf (hunk 5 above)
Comment (paulschwarzenberger) [reply s3/locals.tf]: agreed, now have separate local entries for the log data SNS topic, and the main S3 bucket SNS topic

File: lambda/functions/log-data/lambda_function.py (hunk 2 above)
Comment (TomJKing) [reply lambda/functions/log-data/lambda_function.py:8]: Doh. It was too early in the morning when I looked at this.

File: sns/main.tf (hunk 3 above)
Comment (TomJKing) [reply sns/main.tf:1]: OK. So should the resource be name something more generic, rather than "log_aggregation"?

Comment (paulschwarzenberger) [reply sns/main.tf]: agreed, now implemented

**Pull request merged**